pkg/cli/admin/release/extract_tools.go

```diff
@@ -297,16 +297,123 @@ func (o *ExtractOptions) extractCommand(command string) error {
+
+			To verify the contents of this directory, use the 'gpg' and 'shasum' tools to
+			ensure the archives you have downloaded match those published from this location.
+
+			The openshift-install binary has been preconfigured to install the following release:
+
+			---
+
+		`))
+		if err := describeReleaseInfo(buf, targetRelease.release, false, false, true, false); err != nil {
+			return err
+		}
+		filename := "release.txt"
+		if err := ioutil.WriteFile(filepath.Join(dir, filename), buf.Bytes(), 0644); err != nil {
+			return err
+		}
+		hash := hashFn()
+		hash.Write(buf.Bytes())
+		targetRelease.hashByTargetName[filename] = hex.EncodeToString(hash.Sum(nil))
+	}
+
+	// write a checksum of the tar files to disk as sha256sum.txt.asc
+	if len(targetRelease.hashByTargetName) > 0 {
+		var keys []string
+		for k := range targetRelease.hashByTargetName {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		var lines []string
+		for _, k := range keys {
+			hash := targetRelease.hashByTargetName[k]
+			lines = append(lines, fmt.Sprintf("%s  %s", hash, filepath.Base(k)))
+		}
+		// ensure a trailing newline
+		if len(lines[len(lines)-1]) != 0 {
+			lines = append(lines, "")
+		}
+		// write the content manifest
+		data := []byte(strings.Join(lines, "\n"))
+		filename := "sha256sum.txt"
+		if err := ioutil.WriteFile(filepath.Join(dir, filename), data, 0644); err != nil {
+			return fmt.Errorf("unable to write checksum file: %v", err)
+		}
+		// sign the content manifest
+		if signer != nil {
+			buf := &bytes.Buffer{}
+			if err := openpgp.ArmoredDetachSign(buf, signer, bytes.NewBuffer(data), nil); err != nil {
+				return fmt.Errorf("unable to sign the sha256sum.txt file: %v", err)
+			}
+			if err := ioutil.WriteFile(filepath.Join(dir, filename+".asc"), buf.Bytes(), 0644); err != nil {
+				return fmt.Errorf("unable to write signed manifest: %v", err)
+			}
+		}
+	}
+
+	// if we did not process some targets, report that to the user and error if necessary
+	if len(targetRelease.targetsByName) > 0 {
+		var missing []string
+		for _, target := range targetRelease.targetsByName {
+			missing = append(missing, target.Mapping.From)
+		}
+		sort.Strings(missing)
+		if len(missing) == 1 {
+			return fmt.Errorf("image did not contain %s", missing[0])
+		}
+		return fmt.Errorf("unable to find multiple files: %s", strings.Join(missing, ", "))
+	}
+
+	return nil
+}
+
+type targetReleaseInfo struct {
+	release          *ReleaseInfo
+	willArchive      bool
+	hashByTargetName map[string]string
+	targetsByName    map[string]extractTarget
+}
+
+func (o *ExtractOptions) setReleaseLookup(release *ReleaseInfo, targets []extractTarget, currentOS string, toArchive bool) (targetReleaseInfo, error) {
 	releaseName := release.PreferredName()
 	refExact := release.ImageRef
 	refExact.Ref.Tag = ""
 	refExact.Ref.ID = release.Digest.String()
 	exactReleaseImage := refExact.String()
 
+	tr := targetReleaseInfo{
```

Comment:
More descriptive names, pls, `targetRelease`, for example.

Reply:
updated